### ISSUE TITLE
Migrate stateofus.eth to new ENS registrar

### ISSUE THREAD

**Issue by rachelhamlin** [status-im/status-react#8264]
### Description
*Summary*: `stateofus.eth` must be [migrated to the new ENS registrar](https://medium.com/the-ethereum-name-service/step-by-step-guide-to-migrating-your-eth-name-to-the-new-ens-registrar-5314431e22f2). 

Skeleton issue for now; we should be able to use the [ENS manager](https://manager.ens.domains/) interface to migrate.

**Comment by 3esmit** [comment]
This is a very delicated process, because the domain is held by a contract which dont knows anything about new ENS registrar, and also dont have a default function (needed for migration).
The good news is that ENSUsernames allow migration, and therefore the transfer of the ownership of stateofus.eth to a new address. 

Regarding suggestions from ENS Team, they suggest the ENS domain to be transferred to an EOA to perform safely the migration.
To do that, we will need to execute the migration of the contract to an intermediary disposable contract which only role is to give up ENS Username to an EOA. The EOA would make the necessary transactions and deploy a new ENS Username Registry.

While old users will not be affected, while this migration is happening the service will be unavailable.

Old users will remain under older smart contract terms, where in case of migration, if user wants, funds can be withdrawn by releasing the username without having to wait 1 year.

New registrations will take place in the new deployed contract. 

The deploy of a new smart contract might open possibilities of  new functionalities, updated usernames reserved list, or a completly new contract. 

@corpetty You might be interested in this issue.

**Comment by rachelhamlin** [comment]
Thanks for the information @3esmit.

A few questions so I understand properly:

- Who will own or control the EOA?
- How long might the migration from current contract to disposable to EOA might take, once initiated?
- By `the service will be unavailable`, do you mean no new registration will be possible?

We do not currently have a flow designed for releasing a username. 

This is mandatory under our contract terms, and in my opinion, must then be available before our security audit code freeze. We might have to reprioritise some of our current work to accommodate it.

cc @hesterbruikman @oskarth - I'll set up a call to discuss, would you like to participate?

**Comment by hesterbruikman** [comment]
> cc @hesterbruikman @oskarth - I'll set up a call to discuss, would you like to participate?

Yes, but don't wait on my account. Have a few days off next week and wouldn't want to delay scheduling.

**Comment by oskarth** [comment]
Ditto what Hester said, I'm busy his tweek and next with Swarm summit etc.

**Comment by rachelhamlin** [comment]
Top line from the call we just had:

This is a relatively non-invasive change. With participation from the multisig holders, Ricardo can make the migration happen within 1 week. 

We just need to figure out the security process around it. How extensive of a review will the migrated contract need to undergo? 

Release readiness depends on three things:
- Security process completed
- Contract address changed where referenced in status-react (e.g. new registration flow)
- Simple web DApp available for registrants of the old contract, enabling them to release their name as per the terms & conditions + comms plan about this

@corpetty can you weigh in on the audit process for a change like this one?

**Comment by corpetty** [comment]
audit process is dependent upon the extent of the changes.  @3esmit I need to know exactly how extensive the changes are, and what they are.  If it is more than a little above trivial, it would be worth contracting Sigma Prime to overview the contract changes as they did the original audit. 

If it isn't that much, then we can get ToB to look the changes over once done to sign off on deployment barring any obvious issues.

**Comment by 3esmit** [comment]
I created the contract for migration:
https://github.com/status-im/ens-usernames/blob/new-ens-migrate/contracts/registry/RegistrarMigration.sol

and its test spec:
https://github.com/status-im/ens-usernames/blob/new-ens-migrate/test/registrarmigration.spec.js

The RegistrarMigration.sol is very simple, when it receives the ENS Node from UsernameRegistrar it automatically transfers it to the controller of the domain. 
After this is done, the controller manually calls the migration and sends the ENS domain to a new ENS Usernames contract.

I could handle the process more trustless by instead of transferrign to controller, calling the migration of the ens node, and then automatically migrating and activating the new registry.

**Comment by 3esmit** [comment]
@corpetty

The Migration contract don't require extensive audit https://github.com/status-im/ens-usernames/blob/new-ens-migrate/contracts/registry/RegistrarMigration.sol because it wont hold any assets. Also, the contract is too dumb to be unsafe, it simply does one thing, that is transfer to controller the ens node coming from a specific contract, in this case, the ENS Username Registrar. It also contains a escape hatch `execute function` that allows controller to execute anything in behalf of that contract.

The UsernameRegistrar.sol will use an updated solidity compiler. This changes can be reviewed in this commit https://github.com/status-im/ens-usernames/pull/99/commits/ead468714efa2adf7bebef5ba26b7954f04a18cd#diff-dbff1e6b987cbb9a6b87ea8180c41e72
Notice that all changes in this commit are done to update to the new solidity version. 
And also other small fixes about using new features, such as `.selector`, when possible:
https://github.com/status-im/ens-usernames/commit/6369b587c14a0b8398ccfebff172d23fe85f0c90#diff-dbff1e6b987cbb9a6b87ea8180c41e72
Keep track of the full history here https://github.com/status-im/ens-usernames/commits/new-ens-migrate/contracts/registry/UsernameRegistrar.sol

If we have time to do this change, we can think better what to change, if there is something to. 
The migration process of UsernameRegistrar is the only condition where usernames can be removed/overtaken, this process looks like a "Hard Fork", however we are conducting it behind trust of Status Contributors, trough a Multisig Wallet, which currently controls UsernameRegistrar. The process will look more like a hardfork once a DAO controls the UsernameRegistrar. 

This are the detailed steps to be taken, in case we use a manual migration:

1. Deploy RegistrarMigration.sol init with ENS address, stateofus.eth ensNode and parentRegistry address from a EOA owned by me;
2. Submit a `MultiSigWallet.methods.sendTransaction(UsernameRegistar.address,0, UsernameRegistrar.methods.moveRegistry(RegistrarMigration.address).encodeAbi())`;
3. Wait for approval from more 2 signers of MultiSigWallet;
4. Navigate to https://manager.ens.domains/ and press the button;
5. Deploy new UsernameRegistar.sol and set controller to MultiSigWallet;
6. Transfer ENS domain to new UsernameRegistrar;
7. Submit a `MultiSigWallet.methods.sendTransaction(NewUsernameRegistar.address,0, NewUsernameRegistrar.methods.activate('10000000000000000000').encodeAbi())`;
8. Wait for approval from more 2 signers of MultiSigWallet.

If thats the case I will take over the domain, conduct the migration from my personal EOA and deploy a new UsernameRegistrar.sol.


There is also the option of automating the migration in the Migration contract. This will make a more complex contract to audit, but will be more trustless. I'm working in this version of the contract as well.
1. Deploy the new UsernameRegistrar setting controller as MultiSigWallet;
2. Deploy a RegistrarMigration defining the ENS address, stateofus.eth node hash, oldUsernameRegistrar address, newUsernameRegistrar address;
3. Submit a `MultiSigWallet.methods.sendTransaction(UsernameRegistar.address,0, UsernameRegistrar.methods.moveRegistry(RegistrarMigration.address).encodeAbi())`;
4. Wait for approval from more 2 signers of MultiSigWallet;
5. Submit a `MultiSigWallet.methods.sendTransaction(NewUsernameRegistar.address,0, NewUsernameRegistrar.methods.activate('10000000000000000000').encodeAbi())`;
6. Wait for approval from more 2 signers of MultiSigWallet.


If 
a) the new contract is changed to already have encoded the 10 SNT price, or pass it in constructor, 
b) and assume the contract is live as long it owns the ens node stateofus.eth
= then step ` Submit a MultiSigWallet.methods.sendTransaction(NewUsernameRegistar.address,0, NewUsernameRegistrar.methods.activate('10000000000000000000').encodeAbi())` and its subsequent approval can be removed from the steps.

This all can be tested in Ropsten aswell, before going mainnet.
Ropsten UsernameRegistrar https://ropsten.etherscan.io/address/0x11d9F481effd20D76cEE832559bd9Aca25405841#contracts

**Comment by rachelhamlin** [comment]
Discussed with @3esmit. We are going to make one change to the contract, which is to allow for the list of reserved usernames to be changed only by an operator address. This operator address can be democratically controlled by SNT holders. 

Presently, Status has free reign to alter reserved names at any point.

We need to 1) sort out the level of scrutiny this requires with @corpetty, and 2) determine what is required to switch to the new contract in status-react, and if users require heads up of any down time.

We have until May 4, 2020 to be compliant with ENS' changes.

**Comment by hesterbruikman** [comment]
How is this going @3esmit? I believe @0kok0 was working on a review of the contract changes. Are any updates required as a result of that review?

**Comment by 3esmit** [comment]
We are auditing changes to upgrade the smart contract to a more versatile version, then this migration will happen together with the deploy of new registrar.

**Comment by cammellos** [comment]
Closing this in favor of #11189
